### PR TITLE
JAMES-2544 correct slicing logic testing

### PR DESCRIPTION
**Pull request by chibenwa**
It was relying on UTC, then defeating slicing logic which was no more tested

We keep UTC for slicing agnostic tests in order to keep mail ordering base on enqueued timestamp generated by the clock.

Slicing logic need to override the clock with a fixed timestamp, thus cannot enforce ordering. We relaxed this assertion.

In order to give each tests the clock it needs, we' separated the test classes.

### CODE REVIEW

File: server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueSlicingTest.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+    private static final HashBlobId.Factory BLOB_ID_FACTORY = new HashBlobId.Factory();
+    private static final int THREE_BUCKET_COUNT = 3;
+    private static final int UPDATE_BROWSE_START_PACE = 2;
+    private static final Duration ONE_HOUR_SLICE_WINDOW = Duration.ofHours(1);
+    private static final String SPOOL = "spool";
+    private static final Instant IN_SLICE_1 = Instant.parse("2007-12-03T10:15:30.00Z");
+    private static final Instant IN_SLICE_2 = IN_SLICE_1.plus(1, HOURS);
+    private static final Instant IN_SLICE_3 = IN_SLICE_1.plus(2, HOURS);
+    private static final Instant IN_SLICE_5 = IN_SLICE_1.plus(4, HOURS);
+    private static final Instant IN_SLICE_6 = IN_SLICE_1.plus(6, HOURS);
```
Comment (mbaechler) [inline server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueSlicingTest.java]: it's not IN_SLICE_7 ?

File: server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueSlicingTest.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    void setup(DockerRabbitMQ rabbitMQ, CassandraCluster cassandra) throws Exception {
+        CassandraBlobsDAO blobsDAO = new CassandraBlobsDAO(cassandra.getConf(), CassandraConfiguration.DEFAULT_CONFIGURATION, BLOB_ID_FACTORY);
+        Store<MimeMessage, MimeMessagePartsId> mimeMessageStore = MimeMessageStore.factory(blobsDAO).mimeMessageStore();
+
+        URI amqpUri = new URIBuilder()
+            .setScheme("amqp")
+            .setHost(rabbitMQ.getHostIp())
+            .setPort(rabbitMQ.getPort())
+            .build();
+        URI rabbitManagementUri = new URIBuilder()
```
Comment (mbaechler) [inline server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueSlicingTest.java]: those two URLs could be provided by DockerRabbitMQ

File: server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueSlicingTest.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+            "3-1", "3-2", "3-3", "3-4", "3-5",
+            "5-1", "5-2", "5-3", "5-4", "5-5");
+    }
+
+    @Test
+    void browseAndDequeueShouldCombineWellWhenDifferentSlices() throws Exception {
+        ManageableMailQueue mailQueue = getManageableMailQueue();
+        int emailCount = 5;
+
+        when(clock.instant()).thenReturn(IN_SLICE_1);
```
Comment (mbaechler) [inline server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueSlicingTest.java]: you don't need a mock but a manual clock.
I mean, you could define a ManualClock class that would provide methods like `goForward(Duration)` and `goBackward(Duration)`.

File: server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueSlicingTest.java [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+        Stream<String> names = Iterators.toStream(mailQueue.browse())
+            .map(ManageableMailQueue.MailQueueItemView::getMail)
+            .map(Mail::getName);
+
+        assertThat(names)
+            .containsOnly("3-4", "3-5", "5-1", "5-2", "5-3", "5-4", "5-5");
+    }
+
+    private void enqueueMailsInSlice(int slice, int emailCount) {
```
Comment (mbaechler) [inline server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueSlicingTest.java]: this method is weirdly named : you just enqueue with a given name pattern, you don't specify a slice or a bucket, right ?

Comment (mbaechler) [reply server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueSlicingTest.java]: If I'm correct, you should name this method enqueueMails.
For the naming, you could provide a `Function<Integer, String> name` that you would call with email number. (what you call bucketId for now)

File: server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueSlicingTest.java (hunk 2 above)
Comment (chibenwa) [reply server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueSlicingTest.java]: Handled here https://github.com/linagora/james-project/pull/1757

File: server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/UpdatableTickingClock.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+package org.apache.james.queue.rabbitmq;
+
+import java.time.Clock;
+import java.time.Instant;
+import java.time.ZoneId;
+
+import org.apache.commons.lang3.NotImplementedException;
+
+public class UpdatableTickingClock extends Clock {
```
Comment (mbaechler) [inline server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/UpdatableTickingClock.java]: I'm a bit late, but reviewing Jean's code, I saw that : https://github.com/linagora/james-project/pull/1742/commits/843b66a513c580c562b4bd225b18bdf7e7f2ea64#diff-29254a008b8feddd94407561fedbb9bdR27

File: server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java [hunk 6]
```diff
@@ -88,6 +91,13 @@
 
     private RabbitMQMailQueueFactory mailQueueFactory;
     private UpdatableTickingClock clock;
+    private RabbitMQMailQueue mailQueue;
+
+    @Override
+    public void enQueue(Mail mail) throws MailQueue.MailQueueException {
+        getMailQueue().enQueue(mail);
```
Comment (mbaechler) [inline server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java]: `super.enQueue` ?

File: server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java [hunk 7]
```diff
@@ -231,4 +235,35 @@ private void dequeueMails(int times) {
...
+            throw new NotImplementedException("No timezone attached to this clock");
+        }
+
+        @Override
+        public Clock withZone(ZoneId zone) {
+            throw new NotImplementedException("No timezone attached to this clock");
+        }
+
+        @Override
+        public synchronized Instant instant() {
```
Comment (mbaechler) [inline server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java]: why ?

Comment (chibenwa) [reply server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java]: True, this is no more required on the instant method but on the tick method.

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**